### PR TITLE
Annotation plugin improvements (Fixes #3, #35, #47, #49, #51, #53, #63, #66, and #68)

### LINKED ISSUE #3: Bar chart and box annotation
I'm trying to add a box annotation on a bar chart,
https://jsfiddle.net/jxpx1kup/1/
As you can see this is not working as I expect it. Maybe I'm doing something wrong, but I think that when using bar chart, the function `getPixelForTick` should be used intead of `getPixelForValue` as you can see on this jsfiddle :
https://jsfiddle.net/jxpx1kup/

(The edited line is 53.)

### LINKED ISSUE #49: Check "Scales" options first to continue
Before processing, need to check "Scales" options if it's null or not. Some charts like "Pie chart" etc. may not have "Scales" options.

image

Error line:
https://github.com/chartjs/chartjs-plugin-annotation/blob/master/src/annotation.js#L36,L37

### PR DESCRIPTION
**Pull request by yusufozturk**
Check Scales on ChartOptions to avoid null errors.

Fixes #49

### CODE REVIEW

**Comment by yusufozturk** [comment]
Adding support for non-numeric values on axis

Fixes #47

File: src/annotation.js [hunk 1]
```diff
@@ -95,7 +97,7 @@ module.exports = function(Chart) {
 		afterInit: function(chartInstance) {
 			// Detect and intercept events that happen on an annotation element
 			var watchFor = chartInstance.annotation.options.events;
-			if (watchFor.length > 0) {
+			if (watchFor !== null && typeof watchFor !== "undefined" && watchFor.length > 0) {
```
Comment (compwright) [inline src/annotation.js]: Let's change this to check for an array.

```
if (chartHelpers.isArray(watchFor) && watchFor.length > 0) {
...
}
```

File: src/helpers.js [hunk 2]
```diff
@@ -11,8 +11,17 @@ function objectId() {
 	return Math.random().toString(36).substr(2, 6);
 }
 
-function isValid(num) {
-	return !isNaN(num) && isFinite(num);
+function isValid(rawValue) {
+    // Null and undefined values first
+    if (rawValue === null || typeof (rawValue) === 'undefined') {
+	return NaN;
```
Comment (compwright) [inline src/helpers.js]: This needs to return a boolean.

File: src/helpers.js [hunk 3]
```diff
@@ -11,8 +11,17 @@ function objectId() {
...
+    }
+    // isNaN(object) returns true, so make sure NaN is checking for a number; Discard Infinite values
+    if (typeof (rawValue) === 'number' && !isFinite(rawValue)) {
+	return NaN;
```
Comment (compwright) [inline src/helpers.js]: This needs to return a boolean.

**Comment by yusufozturk** [comment]
Hi Jonathon,
Thanks for your review. I will do changes as soon as possible.
Thanks.

**Comment by JewelsJLF** [comment]
Does this fix include support for category scales? Is it going to be merged into master?

**Comment by compwright** [comment]
@JewelsJLF as you can see in my review it needs a few tweaks before we can merge it. If you're up to the task then feel free to fork and submit a new PR to replace this one.

**Comment by compwright** [comment]
Fixes #3, #35, #47, #49, #51, #53, #63, #66, and #68

**Pull request merged**